plugins/modules/cloud/misc/terraform.py

```diff
@@ -449,12 +474,67 @@ def main():
     if state == 'present' and module.params.get('parallelism') is not None:
         command.append('-parallelism=%d' % module.params.get('parallelism'))
 
-    variables_args = []
-    for k, v in variables.items():
-        variables_args.extend([
-            '-var',
-            '{0}={1}'.format(k, v)
-        ])
+    def process_args(variables, top):
+        variables_args = []
+        lowlevel_out = []
+        for k, v in variables.items():
+            if top:
+                if (isinstance(v, dict)):
+                    variables_args.extend([
+                        "-var",
+                        k + '={' + process_args(v, False) + '}'
+                    ])
+                if (isinstance(v, list)):
+                    l_out = []
+                    for item in v:
+                        l_out.append("{" + process_args(item, False) + "}")
+                    variables_args.extend([
+                        "-var",
+                        k + '=[' + ",".join(l_out) + ']'
+                    ])
+                if ((isinstance(v, int) or isinstance(v, float)) and not isinstance(v, bool)):
+                    variables_args.extend([
+                        "-var",
+                        '{0}={1}'.format(k, v)
+                    ])
+                if (isinstance(v, str)):
+                    variables_args.extend([
+                        "-var",
+                        '{0}={1}'.format(k, v)
+                    ])
+                if (isinstance(v, bool)):
+                    if v:
+                        variables_args.extend([
+                            "-var",
+                            '{0}=true'.format(k)
+                        ])
+                    else:
+                        variables_args.extend([
+                            "-var",
+                            '{0}=false'.format(k)
+                        ])
+
+            else:
+                if (isinstance(v, dict)):
+                    lowlevel_out.append(k + '={' + process_args(v, False) + '}')
+                if (isinstance(v, list)):
+                    lowlevel_out.append(k + '=[' + process_args(v, False) + ']')
+                if ((isinstance(v, int) or isinstance(v, float)) and not isinstance(v, bool)):
+                    lowlevel_out.append('{0}={1}'.format(k, v))
+                if (isinstance(v, str)):
+                    lowlevel_out.append('{0}="{1}"'.format(k, v))
+                if (isinstance(v, bool)):
+                    if v:
+                        lowlevel_out.append('{0}=true'.format(k))
+                    else:
+                        lowlevel_out.append('{0}=false'.format(k))
+        if top:
+            return (variables_args)
+        else:
+            return ",".join(lowlevel_out)
```

Comment:
It looks to me like this whole function can be massively simplified by only computing `lowlevel_out`, and then in case `top` is set converting `lowlevel_out` to `variables_args` by adding a `-var` before every entry. Or do I miss something? (Some special logic is needed for strings, but everything else looks the same to me.)

Reply:
The reason I opted for a function is to make sure nested dictionaries and lists can be processes using recursion. May be there is another way to look at it. I get what you mean, if we isolate the top level `-var` variables, still the remainder can be processed in a recursive function, in that case we may need two different loops. I feel this approach is less code, and easier to track through.

Reply:
Your function can also be written as
```
    def process_args(variables, top=True):
        values = []
        for k, v in variables.items():
            if isinstance(v, dict):
                values.append('%s={%s}' % (k, process_args(v, False)))
            if isinstance(v, list):
                if top:
                    values.append('%s=[%s]' % (k, ','.join(['{%s}' % process_args(item, False) for item in v])))
                else:
                    values.append('%s=[%s]' % (k, process_args(v, False)))
            if isinstance(v, (integer_types, float, bool)):
                values.append('%s=%s' % (k, json.dumps(v)))
            if isinstance(v, string_types):
                if top:
                    values.append('%s=%s' % (k, v))
                else:
                    values.append('%s=%s' % (k, json.dumps(v)))
        if top:
            return [elt for pair in (('-var', item) for item in values) for elt in pair]
        else:
            return ','.join(values)
```
I think this is a lot easier to read.

Please note that your code (and thus also mine) cannot handle lists of anything that's not a dictionary. Is that intentional?

Reply:
> Please note that your code (and thus also mine) cannot handle lists of anything that's not a dictionary. Is that 
> intentional?

You're correct, it's not intentional, I figured it few days back :). The following is a totally legit syntax from terraform:

```
variable "vm_networks" {
  description = "vSphere networks (multiple adapters)"
  type        = list(object(string))
}
```

Which translates to in Asnible: 

```
vm_networks:
 - "asdfg"
 - "qwerty"
```

It was not a total blocker, this still can work with:

```
variable "vm_networks" {
  description = "vSphere networks (multiple adapters)"
  type        = list(object({
    name     = string
  }))
}
```

and 

```
vm_networks:
 - name: "asdfg"
 - name: "qwerty"
```
But let me try to address this use case as well.

Reply:
[b044e96](https://github.com/ansible-collections/community.general/pull/4797/commits/b044e963bff66bb5dd44660a38455370d88bb859)

Restructured the approach to 1) simplify and make it more readable 2) and to fix the above problem.

doing a json.dumps() for strings didn't work for some of the passwords, as terraform was unable to understand escape characters from CLI.

now lists of strings or numbers are possible:

```
variable "vm_networks" {
  description = "vSphere networks (multiple adapters)"
  type        = list(string)
}
```